src/v/archival/ntp_archiver_service.cc

```diff
@@ -581,7 +581,7 @@ static cloud_storage::upload_result process_multiple_upload_results(
 // single future that returns error result of the last failed future or success
 // otherwise.
 static ss::future<cloud_storage::upload_result> aggregate_upload_results(
-  std::vector<ss::future<cloud_storage::upload_result>>&& upl_vec) {
+  std::vector<ss::future<cloud_storage::upload_result>> upl_vec) {
```

Comment:
The way I think of `&&` on a function parameter is way of saying "I _expect_ you to give up ownership of object upon passing it to me", rather than leaving it up to caller's discretion as to whether to move.  Although it's totally unnecessary as the caller can of course std::move into a by-value parameter just as easily.

This feels like it deserves a line in a style guide (along with some of the excellent coroutine gotcha examples)

Reply:
i agree with all this, but i can't help but think you are trying to make a point and i'm missing it entirely. could you clarify?

Reply:
Sorry, that was a bit oblique.  I think I was saying: I'm not sure whether our house style should be to include the && as a hint to callers that you probably want to std::move() into this function, or omit it where it isn't necessary.

Reply:
got it. && certainly has its place, so it's hard to say there is a simple rule, except that its usage as a function parameter should be rare. good call on style guide. i think we started one with @BenPope a while back and then dropped it. was that in github discussions?

> rather than leaving it up to caller's discretion as to whether to move

however, && isn't exactly a forcing function on the outcome. we've had very subtle bugs in the past in which folks incorrectly believe that a && parameter is a sink. the receiver must actively take ownership (in all exit code paths), or the caller will still retain ownership upon return, even though visually it appears otherwise.

Reply:
I'm a bit confused as to whether the discussion is about rvalue references or forwarding references.

Reply:
r-value references

Reply:
But rvalue-reference parameters [are a sink](https://compiler-explorer.com/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:26,endLineNumber:7,positionColumn:26,positionLineNumber:7,selectionStartColumn:26,selectionStartLineNumber:7,startColumn:26,startLineNumber:7),source:'%23include+%3Cmemory%3E%0A%0Aauto+sink(std::unique_ptr%3Cint%3E%26%26+p)+%7B%7D%0A%0Aint+main()+%7B%0A++++auto+val+%3D+std::make_unique%3Cint%3E(42)%3B%0A++++sink(std::move(val))%3B%0A++++return+val+%3F+*val+:+0%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:36.561050241832,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:clang1500,deviceViewOpen:'1',filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!((name:ctre,ver:trunk)),options:'-std%3Dc%2B%2B20+--stdlib%3Dlibc%2B%2B+-O1',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+15.0.0+(Editor+%231)',t:'0'),(h:executor,i:(argsPanelShown:'0',compilationPanelShown:'0',compiler:clang1500,compilerOutShown:'1',execArgs:'',execStdin:'',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-std%3Dc%2B%2B20',source:1,stdinPanelShown:'1',tree:'1',wrap:'1'),l:'5',n:'0',o:'Executor+x86-64+clang+15.0.0+(C%2B%2B,+Editor+%231)',t:'0')),k:36.21269642577051,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+(trunk)',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+clang+15.0.0+(Compiler+%231)',t:'0')),k:27.22625333239749,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

Reply:
Oh, I get it. Yeah, it's a trade-off: Either mistakenly copying an object, or mistakenly not moving from an object.